### PR TITLE
Wheel creation from docker

### PR DESCRIPTION
**Pull request by gusmith**
It may be good not to have it in master, so changing the base from your repo, but I cannot do that from my repo.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
+from setuptools import setup, Extension, find_packages
+import os
+from glob import glob
+
+try:
+    from Cython.Build import cythonize, build_ext
+except ImportError:
+    # No Cython but we may have pre-converted the files
+    maybe_cythonize = lambda x: x
+    cython_cpp_ext = 'cc'
```
Comment (hardbyte) [inline setup.py]: No cython at all so can remove this

File: setup.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+here = os.path.abspath(os.path.dirname(__file__))
+
+requirements = [
+        "bitarray>=0.8.1",
+        "cffi>=1.7",
+        "numpy>=1.14",
+        "mypy-extensions>=0.3",
+        "Cython>=0.29.10"
```
Comment (hardbyte) [inline setup.py]: I think only cffi is required here?

File: DockerfileLinuxWheelCreation [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    git clone --single-branch --branch 3.3.1 https://github.com/microsoft/SEAL.git && cd SEAL/native/src && \
+    cmake . && make install -j8
+RUN apk add --no-cache python3 py3-cffi py3-numpy python3-dev libstdc++ libgomp
+RUN cd /building/secure-p-signature-linkage/ && \
+    python3 -m pip install auditwheel && \
+    python3 setup.py bdist_wheel && \
+    python3 -m auditwheel repair dist/*.whl && \
+    mkdir /result
+
+CMD ["cp", "-r", "/building/secure-p-signature-linkage/dist/", "/result/"]
```
Comment (hardbyte) [inline DockerfileLinuxWheelCreation]: Can we keep just one dockerfile with a build step (build SEAL + compile c code + build wheel) then a "run" stage which demos using the wheel?

File: DockerfileLinuxWheelCreation [hunk 4]
```diff
@@ -0,0 +1,17 @@
+# Stage 1: Building
+FROM alpine:3.10.2 AS build
+# Build tools
+RUN apk add --no-cache git g++ make cmake patchelf
+ADD ./ /building/secure-p-signature-linkage/
+# Build SEAL & spsl
+RUN cd /building && \
+    git clone --single-branch --branch 3.3.1 https://github.com/microsoft/SEAL.git && cd SEAL/native/src && \
+    cmake . && make install -j8
+RUN apk add --no-cache python3 py3-cffi py3-numpy python3-dev libstdc++ libgomp
```
Comment (hardbyte) [inline DockerfileLinuxWheelCreation]: Just confirming we need numpy

File: _cffi_build/seclink.cc [hunk 5]
```diff
@@ -1,8 +1,8 @@
 #include <sstream>
 #include <cassert>
 
-#include "memstream.h"
-#include "seclink_internal.h"
+#include <memstream.h>
```
Comment (hardbyte) [inline _cffi_build/seclink.cc]: revert these

File: _cffi_build/seclink.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+_path = os.path.dirname(__file__)
+sourcefile = os.path.join(_path, 'seclink.cc')
+
+with open(sourcefile, 'r') as f:
+    source = f.read()
+
+
+ffibuilder.set_source(
+    "_truc",
```
Comment (hardbyte) [inline _cffi_build/seclink.py]: Where did this name come from? Maybe `_libseclink`?

File: seclink/sec_link_lib.py [hunk 7]
```diff
@@ -1,59 +1,10 @@
...
-        const char *galkeys, size_t galkeysbytes);
-
-void seclink_decrypt(const seclink_ctx_t ctx,
-        int64_t *outmat, size_t nrows, size_t ncols,
-        const seclink_emat_t inmat,
-        const char *seckey, size_t seckeybytes);
-''')
+from seclink._truc import ffi, lib
 
+lib = ffi.dlopen(None)
```
Comment (hardbyte) [inline seclink/sec_link_lib.py]: We are overriding `lib` here - why?

File: setup.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+#    name="_secure_thingy",
+#    sources=["_cffi_build/seclink.cc", "_cffi_build/encrypt.cc",
+#             "_cffi_build/decrypt.cc", "_cffi_build/multiply.cc",
+#             "_cffi_build/secure-linkage.cc", "_cffi_build/memstream.cc"],
+#    include_dirs=["_cffi_build/"],
+#    language="c++",
+#    extra_compile_args=['-Wall', '-Wextra', '-O3', '-std=gnu++17'],
+#    extra_link_args=["-std=gnu++17"],
+#    define_macros=[('NDEBUG', None)]
+#    )]
```
Comment (hardbyte) [inline setup.py]: Delete?

**Review by hardbyte — CHANGES_REQUESTED** [review]

File: setup.py (hunk 1 above)
Comment (gusmith) [reply setup.py]: :+1:

File: setup.py (hunk 2 above)
Comment (gusmith) [reply setup.py]: Numpy is required too

File: DockerfileLinuxWheelCreation (hunk 3 above)
Comment (gusmith) [reply DockerfileLinuxWheelCreation]: I would prefer not use a single dockerfile, as Hamish has built his dockerfile for a deployment reason.
But I can indeed prepare this dockerfile to show how to use the wheel.

File: DockerfileLinuxWheelCreation (hunk 4 above)
Comment (gusmith) [reply DockerfileLinuxWheelCreation]: Yes, part of the `sec_link_lib.py`.

File: _cffi_build/seclink.cc (hunk 5 above)
Comment (gusmith) [reply _cffi_build/seclink.cc]: :+1:

File: _cffi_build/seclink.py (hunk 6 above)
Comment (gusmith) [reply _cffi_build/seclink.py]: As it was a WIP, mainly to save the changes, I still had the WIP names. "truc" mean "thing" in French. I could have used foo instead, but old habits. `_libseclink` sounds more sensible.

File: setup.py (hunk 8 above)
Comment (gusmith) [reply setup.py]: :+1:

File: seclink/sec_link_lib.py (hunk 7 above)
Comment (gusmith) [reply seclink/sec_link_lib.py]: I was a trial which not removed. :+1:

**Comment by gusmith** [comment]
The new dockerfile is now testing the wheel. It was showing that the previous wheel was actually not working.

File: DockerfileLinuxWheelCreation [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+FROM alpine:3.10.2 AS buildseal
+RUN apk add --no-cache git g++ make cmake patchelf libstdc++ libgomp python3 py3-cffi py3-numpy python3-dev
+# Build SEAL
+RUN mkdir /building && \
+    cd /building && \
+    git clone --single-branch --branch 3.3.1 https://github.com/microsoft/SEAL.git && cd SEAL/native/src && \
+    cmake . && make install -j8
+
+# Stage 2: Build python wheel
+FROM buildseal AS buildwheel
```
Comment (hardbyte) [inline DockerfileLinuxWheelCreation:11]: There isn't any reason to have a stage 2 that just directly keeps building on stage 1? Shouldn't this be a new base and then `COPY` across the built SEAL library

Comment (gusmith) [reply DockerfileLinuxWheelCreation:11]: In stage 1, we are not only building the SEAL library, but also installing it, and al the required python dependencies.
In stage 2, the library is required for building the library and the wheel.
In stage 3, SEAL is still required to run the installed wheel but we do not need the python code, only the wheel created from stage 2.
So we use the same base created from stage 1 (installing SEAL) for the two following stages.

Comment (gusmith) [reply DockerfileLinuxWheelCreation:11]: But it's true we could delete all the SEAL installation process as done in the original dockerfile, but as the goal here is to create a working wheel, I didn't really thin about cleaning too much.

File: seclink/seclink_lib.py [hunk 10]
```diff
@@ -1,59 +1,10 @@
...
-        const char *galkeys, size_t galkeysbytes);
-
-void seclink_decrypt(const seclink_ctx_t ctx,
-        int64_t *outmat, size_t nrows, size_t ncols,
-        const seclink_emat_t inmat,
-        const char *seckey, size_t seckeybytes);
-''')
+from seclink._libseclink import ffi, lib
 
+lib = ffi.dlopen(None)
```
Comment (hardbyte) [inline seclink/seclink_lib.py:7]: ?

File: DockerfileLinuxWheelCreation (hunk 9 above)
Comment (hardbyte) [reply DockerfileLinuxWheelCreation:11]: Ahh okay I follow now

**Pull request merged**